flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/FlinkCalciteCatalogReader.java

```diff
@@ -188,11 +192,46 @@ private static FlinkPreparingTableBase convertCatalogTable(
 			RelDataType rowType,
 			CatalogTable catalogTable,
 			CatalogSchemaTable schemaTable) {
-		return new CatalogSourceTable<>(
-			relOptSchema,
-			names,
-			rowType,
-			schemaTable,
-			catalogTable);
+		if (isLegacyConnectorOptions(catalogTable, schemaTable)) {
+			return new LegacyCatalogSourceTable<>(
+				relOptSchema,
+				names,
+				rowType,
+				schemaTable,
+				catalogTable);
+		} else {
+			return new CatalogSourceTable<>(
+				relOptSchema,
+				names,
+				rowType,
+				schemaTable,
+				catalogTable);
+		}
+	}
+
+	/**
+	 * Checks whether the {@link CatalogTable} uses legacy connector options.
+	 */
+	private static boolean isLegacyConnectorOptions(
+			CatalogTable catalogTable,
+			CatalogSchemaTable schemaTable) {
+		if (catalogTable.getOptions().containsKey(ConnectorDescriptorValidator.CONNECTOR_TYPE)) {
+			return true;
```

Comment:
I think we should consider the case sensitivity here.

Reply:
Why? I checked `TableFactoryService`, we use lower-case as the option key. And `CONNECTOR_TYPE` is lower case.

Reply:
Because `catalogTable.getOptions()` returns the original key.